### PR TITLE
Port BertLayer from Hugging Face transformers

### PR DESCRIPTION
**Pull request by danieldk**
The big one!

I am not sure if we need all the stuff from Hugging Face transformers. Since we are probably only going to finetune, I am not sure if we need `head_mask`. Also not sure if we ever need `attention_mask` or decoder/cross-attention support. For now I have implemented these parts. But we could simplify the implementation later by removing these bits. If you'd prefer the simplicity *now*, I can also remove those bits and do a new PR, which would have as a bonus that we can implement `ModuleT`.

Apologies for the big diff, but the layer needed to be implemented as a whole to allow easy checks against HF transformers output. The next PR will be a lot smaller (the encoder wrapper that applies multiple layers).

### CODE REVIEW

File: src/layers.rs [hunk 1]
```diff
@@ -141,3 +141,60 @@ impl Module for LayerNorm {
...
+        bias: bool,
+    ) -> Self {
+        let vs = vs.borrow();
+
+        let bs = if bias {
+            let bound = 1.0 / (in_features as f64).sqrt();
+            let bs_init = Init::Uniform {
+                lo: -bound,
+                up: bound,
+            };
```
Comment (twuebi) [inline src/layers.rs]: Shouldn't the bias be initialized to zero?

https://github.com/huggingface/transformers/blob/94c99db34cf9074a212c36554fb925c513d70ab1/transformers/modeling_bert.py#L487

Comment (danieldk) [reply src/layers.rs]: I guess so! I was using initialization similar to the PyTorch Linear layer, but I didn't notice that they are overriding this globally.

Comment (twuebi) [reply src/layers.rs]: This override caused some pain when I tried to initialize the bias of the softmax layer to reflect the label distribution.

Comment (danieldk) [reply src/layers.rs]: I can see that, globals are bad. I have now made the initializations arguments to `new` (since the weight initialization was also wrong) and introduced a `bert_linear` function to create a linear layer with BERT's initialization.

File: src/layers.rs [hunk 2]
```diff
@@ -141,3 +141,56 @@ impl Module for LayerNorm {
         )
     }
 }
+
+/// Linear fully-connected layer.
+#[derive(Debug)]
+pub struct Linear {
```
Comment (twuebi) [inline src/layers.rs]: Why implement our own `Linear`? Does `tch::nn::Linear` not offer some things we need?

Comment (danieldk) [reply src/layers.rs]: It does, but it has two downsides:

* We can't directly implement additional methods on it (`from_tensor` in this case), unless we define them as additional traits.
* It expects an `[out_dim, in_dim]` weight matrix and then transposes them during actual matrix multiplication. The BERT matrices are `[in_dim, out_dim]`. Seemed like a bit of a waste to transpose them when initializing the `Linear` layer and then transposing them for each `forward`. Though I guess in the implementation the double transpose results in a normal tensor again, since tansposes are probably a tensor attribute and not an actual transpose.

Comment (danieldk) [reply src/layers.rs]: Relevant issue: https://github.com/pytorch/pytorch/issues/2159

So, `tch-rs` just seems to follow the layout of PyTorch and PyTorch uses a transposed matrix for historical reasons.

Comment (twuebi) [reply src/layers.rs]: I see, wondering whether some of this could go into `tch::nn::Linear`. I guess the transposing is something they won't agree to, PyTorch seems to be doing the same thing. OTOH, not doing the transpose may seem unexpected for someone coming from other torch projects.

`from_tensor` may be nice to have in `tch-rs`, although defining a trait may be difficult since different layers may have a varying number of weights.

Comment (danieldk) [reply src/layers.rs]: Since the transposes have negligible cost, I'll rewrite it to use `Linear`, I think we can replace `from_tensor` by a single trait `PlaceInVarStore`.

Comment (danieldk) [reply src/layers.rs]: `from_tensor` is now replaced by a single `PlaceInVarStore` trait. I've worked around the variable number of parameter tensors by implementing this on the layers rather than the individual weights.

I guess for the future it would be nice to check if we can cheaply find out if a tensor is already in a `VarStore` and making this a no-op when it is.

File: src/bert_model.rs [hunk 3]
```diff
@@ -178,20 +251,578 @@ impl LoadFromHDF5 for BertEmbeddings {
...
+        let attention_scores = attention_scores / (self.attention_head_size as f64).sqrt();
+        let attention_scores = match attention_mask {
+            Some(mask) => attention_scores + mask,
+            None => attention_scores,
+        };
+
+        // Convert the raw attention scores into a probability distribution.
+        let attention_probs = attention_scores.softmax(-1, Kind::Float);
+
+        // Drop out entire tokens to attent to, following the original
```
Comment (twuebi) [inline src/bert_model.rs]: attend

Comment (danieldk) [reply src/bert_model.rs]: Thanks, fixed.

**Comment by twuebi** [comment]
Not sure whether we need the cross attention stuff etc. The only use case without also implementing the data part is library use. So the question is: do we want to be a general hf transformers port? If yes, then we should probably factor the transformers stuff out into its own crate, otherwise, I don't think there's a need to keep all features.

**Comment by danieldk** [comment]
> Not sure whether we need the cross attention stuff etc. The only use case without also implementing the data part is library use. So the question is: do we want to be a general hf transformers port? If yes, then we should probably factor the transformers stuff out into its own crate, otherwise, I don't think there's a need to keep all features.

I don't think we want to be a general port. If people want to train fresh BERT/RoBERTa/XLNET, etc. models, they could just use HF transformers. I don't think there is a benefit of doing that in Rust. Dropping the training-from-scratch stuff also makes it easier/quicker to implement additional models, without maintenance becoming a full-time task.

I will close this PR and open a new PR with those features removed.

**Comment by twuebi** [comment]
Not sure what the cross attention stuff is used for, maybe for multiple-choice QA? IMO, we could drop it.
